lib/bulkhead/Bulkhead.ts

```diff
@@ -0,0 +1,54 @@
+import { DEFAULT_SIZE } from './BulkheadOptions';
+
+export class Bulkhead {
+
+  private inFlight: number = 0;
+  private queue: any[] = [];
+
+  constructor(
+    private readonly threshold: number,
+    private readonly size: number = DEFAULT_SIZE,
+  ) { }
+
+  public pass(): boolean {
+    return !this.size || this.queue.length < this.size;
+  }
+
+  public run(scope: any, method: () => Promise<any>, args: any) {
+    if (this.inFlight < this.threshold) {
+      return this.runInstant(scope, method, args);
+    }
+
+    return this.addToQueue(scope, method, args);
+  }
+
+  private async runInstant(scope, method: () => Promise<any>, args: any) {
+    let response;
+    this.inFlight += 1;
+
+    try {
+      response = await method.apply(scope, args);
+    } catch (e) { }
```

Comment:
mmm... seems like if my method throws, the error will be swallowed and an `undefined` value returned.